appinventor/components/src/com/google/appinventor/components/runtime/PieChartDataModel.java

```diff
@@ -0,0 +1,221 @@
+      // 2-tuples are invalid when null entries are present, or if
+      // the number of entries is not sufficient to form a pair.
+      // TODO: Show toast error notification
+    }
+
+    return null;
+  }
+
+  @Override
+  public YailList getTupleFromEntry(Entry entry) {
+    // Cast Entry to PieEntry (safe cast)
+    PieEntry pieEntry = (PieEntry) entry;
+
+    // Create a list with the X and Y values of the entry, and
+    // convert the generic List to a YailList
+    List tupleEntries = Arrays.asList(pieEntry.getLabel(), pieEntry.getY());
+    return YailList.makeList(tupleEntries);
+  }
+
+  @Override
+  protected void setDefaultStylingProperties() {
+    // Set spacing between each slice
+    dataset.setSliceSpace(3);
+  }
+
+  @Override
+  protected YailList getDefaultValues(int size) {
+    // Default values for PieChartBaseDataModel should be
+    // integers from 0 to N (0, 1, 2, ...)
+    // TODO: This could be updated in the future to return Strings
+    // TODO: such as "Entry 1", "Entry 2", ... for x and 1,2,3,...,N for y
+    ArrayList<Integer> defaultValues = new ArrayList<>();
+
+    for (int i = 0; i < size; ++i) {
+      defaultValues.add(i);
+    }
+
+    return YailList.makeList(defaultValues);
+  }
+
+  /**
+   * Sets the colors of the Data Series from the specified
+   * colors YailList.
+   *
+   * Invalid (non-color) entries are simply ignored.
+   * @param colors YailList of entries expected to contain colors (integers)
+   */
+  public void setColors(YailList colors) {
+    // Parse the entries of the YailList
+    List<Integer> resultColors = new ArrayList<Integer>();
+
+    for (int i = 0; i < colors.size(); ++i) {
+      // Get the element of the YailList as a String
+      String color = colors.getString(i);
+
+      try {
+        // Parse the color value and add it to the results List
+        int colorValue = Integer.parseInt(color);
+        resultColors.add(colorValue);
+      } catch (NumberFormatException e) {
+        // Skip invalid entry
+      }
+    }
+
+    // Set the colors to the dataset
+    getDataset().setColors(resultColors);
+
+    // Update the legend
+    for (int i = 0; i < legendEntries.size(); ++i) {
+      // Since the number of colors might be less, the index
+      // of the color to use should be modulo the color count
+      int index = i % getDataset().getColors().size();
+
+      // Set the corresponding color to the Legend entry (since
+      // LegendEntry is referenced, this will automatically update
+      // the central Legend that contains all the entries, and not
+      // just the local ones)
+      legendEntries.get(i).formColor = getDataset().getColors().get(index);
+    }
+  }
+
+  @Override
+  protected boolean areEntriesEqual(Entry e1, Entry e2) {
+    if (!(e1 instanceof PieEntry && e2 instanceof PieEntry)) {
+      return false;
+    }
+
+    PieEntry p1 = (PieEntry) e1;
+    PieEntry p2 = (PieEntry) e2;
+
+    return p1.getLabel().equals(p2.getLabel())
```

Comment:
Test cases:
- [ ] y values equal, x values equal
- [ ] y values not equal, x values equal
- [ ] y values equal, x values not equal